PruManager RProc+UIO implementation

DhruvaG2000 opened the pull request:
The code has been tested on the BELA image so far using the `sample_streamer` example.
- I am able to get audio output fine.
- So far I have ported all the prussdrv functions mentioned [here on forum](https://forum.beagleboard.org/t/bela-support-for-bbai-later-ti-chips/29257/10?u=dhruvag2000) into PruManager.
- If this PR receives your approval(ie. if the code looks okay to you) then I can move on to work on debugging `Rproc` using newer TI 4.19 kernels, as UIO is possibly **not properly working** even on the `BBB v4.19`. I planned to do this in a step by step approach first where I first start by replacing only the `start stop` using rproc and let uio handle the rest. However in the light of the fact that UIO drivers aren't working properly, I will need to completely adapt to RProc + Mmap approach in one go (which can take upto a week or two)

giuliomoro committed:
pru: use McaspIRQ code for all boards. pru/pru_rtaudio.p now becomes unused

DhruvaG2000 committed:
core/PRU.cpp: tested PruManagerUIO::stop()

DhruvaG2000 committed:
PRU: moved copy and load-fw() in start() in rprocmap and also moved a few headers from .h to .cpp

DhruvaG2000 committed:
PruManager: moved copy and load-fw() in start()   and also moved a few headers from .h to .cpp

DhruvaG2000 committed:
Merge branch 'multi-codec' into PruManager

DhruvaG2000 committed:
added address of prumanageruio object in prumem constructor

DhruvaG2000 committed:
implemented UIO start

DhruvaG2000 committed:
PRU: replaced prussdrv_map_prumem with UIO class function

DhruvaG2000 committed:
PRU.cpp: prumanager = new PruManagerUio moved into constructor PRU

DhruvaG2000 committed:
PRU: add flag ENABLE_PRU_UIO

- Added ENABLE_PRU_UIO Flag to Makefile
- The new UIO Flag has been added as a macro to:
    - `core/PRU.cpp`
    - `core/PruManager.cpp`
    - include/PruManager.h
- also fix data ram line which was earlier using shared ram.

DhruvaG2000 committed:
PruManager: update mapping approach

- This commit implements major changes like replacing map_pru_mem with 2
seperate functions getSharedMemory and getOwnMemory
- added many conditional macros to help define according to requirement
to use rproc or uio.
- added an ifdef before the exec code part. Not sure if I should make
another function instead for execution under base class, not sure if
that is necessary because at the moment the loading code and executing
is handled in rproc::start()
- Files modified: core/PRU.cpp, core/PruManager.cpp, include/PruManager.h

DhruvaG2000 committed:
PRU: Replaced spaces with tabs

- modified:   core/PruManager.cpp
- modified:   include/PruManager.h

DhruvaG2000 committed:
core/PRU.cpp: pass parameters to constructor

pass pru_number and verbose flags to respective constructors when
prumanager object is created in PRU constructor

DhruvaG2000 committed:
include/PruManager.h: created Mmap objs

DhruvaG2000 committed:
Syntax indentation improvements

modified:   Makefile
modified:   core/PRU.cpp

DhruvaG2000 committed:
Commit changes that were suggested by mentor in meet

modified:   Makefile
modified:   core/PRU.cpp
modified:   core/PruManager.cpp
modified:   include/PRU.h
modified:   include/PruManager.h

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,78 @@
+	Mmap sharedMemory;
+	long mem2;
+};
+#endif	// end condition for ENABLE_PRU_RPROC
```
should be 

```
#endif // ENABLE_PRU_RPROC
```

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,78 @@
+#if ENABLE_PRU_RPROC == 1
+class PruManagerRprocMmap : public PruManager
+{
+/* use rproc for start/stop and mmap for memory sharing
```
one line comment, please use `//`

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,78 @@
+/* 
+ * PruManager.h
+ *
+ * includes the classes required for introducing RProc functionality
```
a more appropriate description would be "support for interaction with PRU via rproc/mmap and/or uio/libprussdrv"

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,78 @@
+	unsigned int pru_num, verbose;
+	std::string state;
+	PruManager();
+	virtual void readstate() = 0;
```
I don't think it's a good design decision to have `state` as a public member  variable and a `void readstate()` method. I'd much prefer:
```
typedef enum { kRunning, KState} State;
virtual State getState();
```

DhruvaG2000 replied:
> I don't think it's a good design decision to have state as a public member variable

Oh well the reason for that is that UIO and RPRoc have different ways of reading state. in UIO there isn't exactly a direct method, hence we had discussed in meet to maintain a certain local variable that gets toggled when UIO start is called.
And then for rproc we just read the contents of "state"... 
So since these both have different approaches, but to keep the same function name we had made it virtual in the base class. The global variable is because if it were private then the child would not inherit it and so would not be able to read it if needed anywhere.

giuliomoro replied:
it should be `protected` (instead of `public` or `private`) for the child to be able to access it. Make it a `typedef enum`, no point in having it a string.

DhruvaG2000 replied:
> Make it a typedef

okay if protected, then it can be done. Then I will just rename `readstate` to `getState` and keep the function implementations the same, just change the string part to the use new `typedef`

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,157 @@
+{	//performs echo stop > state
+	if(verbose)
+		std::cout << "Stopping the PRU1_0 \n";
+	//mode = TRUNCATE; by default
```
not needed

giuliomoro replied:
not needed, here and elsewhere

giuliomoro commented on Makefile:
```diff
@@ -294,7 +294,18 @@ ifeq ($(XENOMAI_VERSION),3)
+endif
+
+#	Flag for using/ not using UIO. Set it to 0 to use RPROC instead
+ENABLE_PRU_UIO = 1
```
these flags could be set automatically based on `IS_AM572x`

DhruvaG2000 committed:
Fix comments as suggested in review Part 1

modified:   core/PruManager.cpp
modified:   include/PruManager.h

DhruvaG2000 committed:
Makefile: flags set automatically based on IS_AM572x

giuliomoro commented on Makefile:
```diff
@@ -294,7 +294,22 @@ ifeq ($(XENOMAI_VERSION),3)
 endif
 
-DEFAULT_COMMON_FLAGS := $(DEFAULT_XENOMAI_CFLAGS) -O3 -g -march=armv7-a -mtune=cortex-a8 -mfloat-abi=hard -mfpu=neon -ftree-vectorize -ffast-math -DNDEBUG -D$(BELA_USE_DEFINE) -I$(BASE_DIR)/resources/$(DEBIAN_VERSION)/include -save-temps=obj
+IS_AM572x = $(shell grep -q AI /proc/device-tree/model && echo '-DIS_AM572x' || echo ' ')
```
you shouldn't need the `|| echo ' '` because when using the variable you will always surround it with whitespaces.

DhruvaG2000 replied:
> when using the variable you will always surround it with whitespaces

Oh I see.. then I will remove the `|| ...` part.

giuliomoro commented on core/PRU.cpp:
```diff
@@ -770,6 +778,10 @@ int PRU::start(char * const filename, const McaspRegisters& mcaspRegisters)
 			return 1;
 		}
 	}
+#endif
```
`#endif // ENABLE_PRU_UIO`

giuliomoro commented on core/PRU.cpp:
```diff
@@ -222,6 +226,12 @@ PRU::PRU(InternalBelaContext *input_context)
   audio_expander_filter_coeff(0), pruUsesMcaspIrq(false), belaHw(BelaHw_NoHw)
 {
+#if ENABLE_PRU_UIO == 1
+	prumanager = new PruManagerUio(pru_number, gRTAudioVerbose);
```
use camel case for variable names: `pruManager`

giuliomoro commented on core/PRU.cpp:
```diff
@@ -384,12 +394,9 @@ int PRU::initialise(BelaHw newBelaHw, int pru_num, bool uniformSampleRate, int m
-	}
-	pruMemory = new PruMemory(pru_number, context);
+
+	prumanager->start();
```
`prussdrv_open()` should be replaced by  ~`new PruMemory()`~ `new PruManager()`  and  `prussdrv_exec_code` should be replaced by `pruManager->start()`.

DhruvaG2000 replied:
>prussdrv_open() should be replaced by new PruMemory

Are you sure? I think what `prussdrv_open()` partially does is checks if it "failed to open PRU driver" and accordingly returns 1 or 0 and then accordingly `if-else` works. However we would be eliminating the return _inside_ `if-else` it if we were to be placed it in the constructor, is that okay?

giuliomoro replied:
sorry, it should have been `new PruManager`

giuliomoro commented on core/PRU.cpp:
```diff
@@ -770,6 +778,10 @@ int PRU::start(char * const filename, const McaspRegisters& mcaspRegisters)
 	}
+#endif
+#if ENABLE_PRU_RPROC == 1
+// do something else probably? Or simply not required.
```
This is where we should be calling  `pruManager->start()`.

I guess the issue here is how to handle `prussdrv`'s flexibility of executing embedded vs external binary code. I suggest the following:
- add a `std::string` argument:  `PruManager::start(const std::string& path)`.
- if (`"" == path`), `Uio` will call `prussdrv_exec_code()`, while `Rproc` will use the default path of the binary built by the `pasm/clpru` workflow.
- if a non-empty string is passed, `Uio` will call `prussdrv_exec_program()`, while `Rproc` will use the provided path, which is assumed to be that of an externally-generated file that was built with the `pasm/clpru` workflow.

One issue remains here, that of what happens for `pru_rtaudio.p` vs  `pru_rtaudio_irq.p`, but given how we are hoping to phase out the former soon, we could just hardcode either for now.

The code here would then look something like:

```
	if(filename[0] == '\0') { //if the string is empty, load the embedded code
		if(gRTAudioVerbose)
			printf("Using embedded PRU code\n");
		if(pruManager->start("")) {
			fprintf(stderr, "Failed to execute PRU code\n");
			return 1;
		}
	} else {
		if(gRTAudioVerbose)
			printf("Using PRU code from %s\n",filename);
		if(!pruManager->start(filename)) {
			fprintf(stderr, "Failed to execute PRU code from %s\n", filename);
			return 1;
		}
	}
```
which could be further simplified to:
	
```
	if(gRTAudioVerbose)
		printf("Using %s PRU code\n", 0 == strlen(filename) ? "embedded" : filename);
	if(!pruManager->start(filename)) {
		fprintf(stderr, "Failed to execute PRU code\n");
		return 1;
	}
```

DhruvaG2000 replied:
(edited once)
1. `prussdrv_exec_code(pru_number, pruCode, pruCodeSize))` has 3 input parameters unlike `prussdrv_exec_program(pru_number, filename))` which has only 2. So should I do function overloading by defining 2 different `start()` functions then? (To be noted, the overloading will prove to be useless in rproc and one of the overloaded function will have a blank definition.)
2. Or, as [you said earlier in this comment,](https://github.com/giuliomoro/Bela-dhruva/pull/1#discussion_r672162925) we could just put the  `prussdrv_exec_code()` part in the class implementation and leave [exec_program()](https://github.com/giuliomoro/Bela-dhruva/blob/multi-codec/core/PRU.cpp#L766-L770) as is. Again, this wouldn't be ideal because on non-uniformity in the code.
3. **Current implementation:** Since we can just hardcode either for now, then I can `include "PruBinary.h"` in `PruManager.cpp` and this way we can have the values of `pruCode, pruCodeSize` right in the start function itself. (downside is that this will be redundant computation, which is sort of a bad practice)

giuliomoro replied:
>  as you said earlier in this comment, we could just put the prussdrv_exec_code() part in the class implementation and leave exec_program() as is.

that's not quite what I meant.

> Since we can just hardcode either for now, then I can include "PruBinary.h" in PruManager.cpp and this way we can have the values of pruCode, pruCodeSize right in the start function itself

that's more like what I meant.

DhruvaG2000 replied:
> that's more like what I meant

Okay, please check the latest commit: https://github.com/giuliomoro/Bela-dhruva/pull/1/commits/a450efa930d325e0172b2c39ef2bf7aa6442618f , if it seems okay to you then I will remove the residual comments from the code.

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,153 @@
+	statePath = basePath + "state";
+	firmwarePath = basePath + "firmware";
+	firmware = "am57xx-pru" + std::to_string(pruss) + "_" + std::to_string(prucore) + "-fw";
+	firmwareCopyCommand = "sudo rm /lib/firmware/" + firmware + ";sudo ln -s /home/debian/Bela-dhruva/pru/blinkR30.pru1_0.out /lib/firmware/" + firmware; // **NOTE**: Change the name of .out file according afterwards here and below cases.
```
build this string programmatically, do not hardcode the `/home/debian/Bela-dhruva/...` path. Maybe set a define in the `Makefile` for what this path should be. I think you can use `ln -s -f` instead  of `rm` + `ln`.

DhruvaG2000 replied:
>build this string programmatically
umm I definitely could, but is that really necessary? There is always only going to be that `pru_rt_audio_irq.p` file right? This is currently done only at a testing stage. Maybe next couple of days I will see if running some example the rproc part is atleast able to flash the PRU correctly or not.

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,153 @@
+{
+	// prussdrv_init() I guess.
+	// prussdrv_exec_program(pru_num, *filename);	// Maybe define filename in constructor
+	prussdrv_init();
```
`prussdrv_init` and `prussdrv_open` should be in the constructor.

DhruvaG2000 committed:
Remove ReadState functions and variables

modified:   core/PruManager.cpp
modified:   include/PruManager.h

DhruvaG2000 committed:
Makefile: Replace tabs with spaces

DhruvaG2000 committed:
Minor fixes

modified:   core/PruManager.cpp
modified:   include/PruManager.h

DhruvaG2000 committed:
Use Camel case for pruManager

modified:   core/PRU.cpp
modified:   include/PRU.h

DhruvaG2000 committed:
modified:   core/PRU.cpp

add comments in front of endif's

DhruvaG2000 committed:
PruManager: Move contents from start() -> constructor

modified:   core/PRU.cpp
modified:   core/PruManager.cpp

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,149 @@
+
+}
+
+int PruManagerUio::start(const char* path)
```
should return 0 on success

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,149 @@
+void PruManagerRprocMmap::stop()
+{	//performs echo stop > state
+	if(verbose)
+		std::cout << "Stopping the PRU1_0 \n";
```
you should build the printed string programmatically based on actual PRU number

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,149 @@
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(0,0x4a334000));
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(1,0x4a338000));
+#endif
+	long addr2 = 0x4b200000;
```
this is a local variable. Remove `long` and add it as a member variable.

giuliomoro replied:
also, its assignment should also be part of the `#ifdef`.

giuliomoro replied:
(unless it's the same for AM572x and AM335x)

giuliomoro replied:
I feel like we discussed this already ...

also, rename it from `addr2` to something more meaningful.

DhruvaG2000 replied:
Yes, this is still in my to-do list...

DhruvaG2000 replied:
done

DhruvaG2000 commented:
> I'd prefer `const std::string&`

Well but the prussdrv function accepts a char* as argument so to avoid `str.c_str();` I used this. But if you prefer that, then I will use string.

DhruvaG2000 committed:
PruManager.cpp: Built start/stop PRU string

DhruvaG2000 committed:
PruManager.cpp: uio::start() return 0 on success

DhruvaG2000 committed:
Pass firmware from Makefile

modified:   Makefile
modified:   core/PRU.cpp
modified:   core/PruManager.cpp

DhruvaG2000 committed:
PRU.cpp: remove commented code from constructor

DhruvaG2000 committed:
PruManager: changed path from char to string

modified:   core/PruManager.cpp
modified:   include/PruManager.h

giuliomoro commented on core/PRU.cpp:
```diff
@@ -384,12 +388,15 @@ int PRU::initialise(BelaHw newBelaHw, int pru_num, bool uniformSampleRate, int m
+	pruManager = new PruManagerRprocMmap(pru_number, gRTAudioVerbose);
+#endif	// ENABLE_PRU_UIO
+
+	// pruManager->start();
```
remove comment and empty line above it

giuliomoro commented on core/PRU.cpp:
```diff
@@ -1514,15 +1518,15 @@ void PRU::waitForFinish()
 {
     /* Disable PRU and close memory mapping*/
-    prussdrv_pru_disable(pru_number);
+    pruManager->stop();
```
while you are at it, please replace the leading spaces in this and the above line with a tab.

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	firmware = "am57xx-pru" + std::to_string(pruss) + "_" + std::to_string(prucore) + "-fw";
+# ifdef firmwareBelaRProc
+	std::string firmwareBela = firmwareBelaRProc;	// Incoming string from Makefile
+# endif	// firmwareBelaRProc
```
as written, if `firmwareBelaRProc` is not defined, you will be left  with a confusing compiler error at the next line. If the definition of `firmwareBelaRProc` is required, then add an explicit
```
#else // firmwareBelaRProc
#error (meaningful error description here)
#endif // firmwareBelaRProc
```

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+void PruManagerRprocMmap::stop()
+{	// performs echo stop > state
+	if(verbose)
+		std::cout << "Stopping the PRU" << std::to_string(pruss) + "_" + std::to_string(prucore) << "\n";
```
If you can, please use `printf()` instead of `std::cout`. That's what most of the core code does (or should do!). But don't worry about it too much.

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	stop();
+	system(firmwareCopyCommand.c_str());	// copies fw to /lib/am57xx-fw
+	if(verbose)
+		std::cout << "Loading firmware into the PRU" << std::to_string(pruss) + "_" + std::to_string(prucore) << "\n";
```
in this case for instance you clearly see the sheer power and elegance of `printf()` against `std::cout`.

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	// 1 : pru-core 1 in BBB -> 4a338000
+
+# ifdef IS_AM572x	// base addresses for BBAI
+	pru_addr.insert(std::pair<unsigned int, unsigned int>(0,0x4b234000));
```
space after `0,`

giuliomoro commented on include/PRU.h:
```diff
@@ -178,6 +179,8 @@ class PRU
 	// Exit the whole PRU subsystem
 	void exitPRUSS();
 
```
this empty line can also go

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	else
+		return pruDataRam;
+}
+
```
no need for two empty lines. One is enough

giuliomoro commented on core/PRU.cpp:
```diff
@@ -755,22 +762,19 @@ int PRU::start(char * const filename, const McaspRegisters& mcaspRegisters)
-			fprintf(stderr, "Failed to execute PRU code from %s\n", filename);
-			return 1;
-		}
+#if ENABLE_PRU_UIO == 1
```
at this point this `#ifdef` should  no longer be needed.

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,74 @@
+	void* getOwnMemory();
+	void* getSharedMemory();
+private:
+	std::map<unsigned int, unsigned int> pru_addr;	// prunum : pru address
```
rename `pru_addr` to `pruRamAddr" and `pruss_addr` to `sharedRamAddr`

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	//
+	// 0 : pru-core 0 in BBB -> 4a334000
+	// 1 : pru-core 1 in BBB -> 4a338000
+
```
one further thought, probably a better way to write this code is to have:
```
// in class declaration
std::vector<uint32_t> prussAddresses;
// in class declaration or as constants at the top of the cpp file
const std::vector<uint32_t> prussOwnRamOffsets = {0x0, 0x2000};
const uint32_t prussSharedRamOffset = 0x10000;

// in constructor
#ifdef IS_AM572x
prussAddresses.push_back(0x4b200000);
prussAddresses.push_back(0x4b280000);
#else // IS_AM572x
...
#endif // IS_AM572x
...
// in getOwnMemory()
	return ownMemory.map(prussAddresses[pruss] + prussOwnRamOffsets[prucore], 0x2000);
// in getSharedMemory()
	return sharedMemory.map(prussAddresses[pruss] + prussSharedRamOffset, 0x8000);

```

DhruvaG2000 replied:
> const std::vector<uint32_t> prussOwnRamOffsets = {0x0, 0x2000};

You probably mean  `{0x4000, 0x8000};`? Since it's `0x4b234000` and `0x4b238000` for PRU1_0, PRU1_1 which is `prussAddresses[0] // =0x4b200000` + `0x4000` and + `0x8000` respectively.

-- edit -- 
Going thru TRM once again to be sure

DhruvaG2000 replied:
done in 2f8a6fcca96acc644a00bc40a58db96dbec3bf9f

giuliomoro commented on core/PruManager.cpp:
```diff
@@ -0,0 +1,156 @@
+	IoUtils::writeTextFile(statePath, "stop");
+}
+
+int PruManagerRprocMmap::start(const std::string& path)
```
you should use the `path` argument as per  https://github.com/giuliomoro/Bela-dhruva/pull/1/files#r672174411

DhruvaG2000 committed:
Makefile: Add PruManager to CORE_CORE_OBJS

DhruvaG2000 committed:
PRU.cpp: Cleanups and indent fixes

DhruvaG2000 committed:
PruManager.cpp: Add error message if firmware file not defined

DhruvaG2000 committed:
delete excess blank lines

DhruvaG2000 committed:
PruManager: rename variables

rename pru_addr to pruRamAddr and pruss_addr to sharedRamAddr
modified:   core/PruManager.cpp
modified:   include/PruManager.h

DhruvaG2000 committed:
PruManager.cpp: shorten include path to PruMan..h

giuliomoro commented on core/PRU.cpp:
```diff
@@ -15,7 +15,11 @@
 
 #include "../include/PRU.h"
 #include "../include/PruBinary.h"
```
this should no longer be needed here, then?

giuliomoro replied:
similarly, pruCode, pruCodeSize should go

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,74 @@
+{
+// use rproc for start/stop and mmap for memory sharing
+public:
+	PruManagerRprocMmap(unsigned int pruNum = 0, unsigned int v = 0);
```
add a virtual destructor (compiler is complaining about it)

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,74 @@
+{
+// expose parameters for the relevant paths
+public:
+	unsigned int pru_num, verbose;
```
add a virtual destructor (compiler is complaining about the lack of it)

DhruvaG2000 replied:
done

giuliomoro commented on include/PruManager.h:
```diff
@@ -0,0 +1,74 @@
+{
+// expose parameters for the relevant paths
+public:
+	unsigned int pru_num, verbose;
```
the above variables shouldn't be public

DhruvaG2000 replied:
done 
4eae814

giuliomoro committed:
PruManager: can load embedded code based on pruUsesMcaspIrq or arbitrary code.

DhruvaG2000 committed:
bela_hw_settings.h: Add conditional comiplation feature for I2c

It should now support I2C on both the BBB and the BBAI

DhruvaG2000 committed:
core/Mcasp.cpp: add conditional compile for auxClkIn frequency

DhruvaG2000 committed:
core/PRU.cpp: solve ret variable causing not declared compile error

DhruvaG2000 committed:
PruManager: update AI PRU addresses

and also add virtual destructor

modified:   core/PruManager.cpp
modified:   include/PruManager.h

DhruvaG2000 committed:
Revert "pru: use McaspIRQ code for all boards. pru/pru_rtaudio.p now becomes unused"

This reverts commit bf14f5172e25b0f3b1f9fc0b7ef573b8fa9d4786.

DhruvaG2000 committed:
Create rproc-build and copy ai-rtaudio code into pru/.

new file:   pru/ai-pru_rtaudio.p
new file:   resources/rproc-build/common/Makefile
new file:   resources/rproc-build/common/README.md
new file:   resources/rproc-build/common/am335x_pru.cmd
new file:   resources/rproc-build/common/am57xx_pru.cmd
new file:   resources/rproc-build/common/init_pins_empty.h
new file:   resources/rproc-build/common/prugpio.h
new file:   resources/rproc-build/common/resource_table_0.h
new file:   resources/rproc-build/common/resource_table_1.h
new file:   resources/rproc-build/common/resource_table_empty.h
new file:   resources/rproc-build/common/write_init_pins.sh
new file:   resources/rproc-build/template.pru1_1.c

DhruvaG2000 committed:
pru/ai-pru_rtaudio.p: conditional compile using IS_AM57x flag

DhruvaG2000 committed:
pru/ai-pru_rtaudio.p: Remove space after # that causes error

DhruvaG2000 committed:
pru_rtaudio.p: Add conditional compiling

deleted:    pru/ai-pru_rtaudio.p
modified:   pru/pru_rtaudio.p

DhruvaG2000 committed:
deleted:    resources/rproc-build/common/Makefile

DhruvaG2000 committed:
PruManager: replace map with vector approach for RProc+Mmap addresses

DhruvaG2000 committed:
include/PruManager.h: make pru_num, verbose private variables

DhruvaG2000 committed:
Mcasp.cpp: add endif comment

DhruvaG2000 committed:
PRU.cpp: fix endif comment

DhruvaG2000 committed:
PruManager.cpp: minor fix, declared prussSharedRamOffset and prussOwnRamOffsets

DhruvaG2000 committed:
Merge branch 'PruManager' of github.com:giuliomoro/Bela-dhruva into PruManager

giuliomoro committed:
Makefile: simplified checks for IS_AM572x. Distinguished between Rproc/Uio/board. Removed dependency on libprussdrv when using rproc

giuliomoro committed:
Makefile: BELA_USE_DEFINE is now ?= so it can be overridden more easily in CustomMakefileTop.in

giuliomoro committed:
Makefile: exclude PruBinary when not needed

giuliomoro committed:
Makefile: build .out file via pasm+clpru when ENABLE_PRU_RPROC==1. Tidied up related and unrelated rproc-build files

giuliomoro committed:
PruManager: rationalised firmwareCopyCommand and related prinouts

giuliomoro committed:
PruManager: properly warn for untested code path

giuliomoro committed:
PruManager: removed unused variables, refactored printing, delegated some variables to the base class, initialised by the base constructor

giuliomoro committed:
PruManager: NIT: camelCased pruCore

giuliomoro committed:
pru: enable mcasp module clock from PRU during initialisation

giuliomoro committed:
pru: wait for a few cycles after enabling McASP module clock in order to ensure proper initialisation

giuliomoro commented:
rproc + uio stuff refactored and (until 551dd96d5a0dabe229923c3923122e3c3cb4e73c)  merged into multi-codec
